common/timer.go

```diff
@@ -0,0 +1,206 @@
+package common
+
+import (
+	"math"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/common/types"
+)
+
+// TimerStore manages timers to efficiently support future timeouts. Timeouts
+// can be based on either block-height or block-timestamp. When a timeout occurs,
+// a designated callback function is called with the details (ctx and data).
+//
+// Once instantiated with NewFixationStore(), it offers the following methods:
+//    - WithCallbackByBlockHeight(callback): sets the callback for block-height timers
+//    - WithCallbackByBlockTime(callback): sets the callback for block-time timers
+//    - AddTimerByBlockHeight(ctx, block, data): add a timer to expire at block height
+//    - AddTimerByBlockTime(ctx, timestamp, data): add timer to expire at block timestamp
+//    - Tick(ctx): advance the timer to the ctx's block (height and timestamp)
+//
+// How does it work? The explanation below illustrates how the data is stored, assuming
+// the user is the module "package":
+//
+// 1. When instantiated, TimerStore gets a `prefix string` - used as a namespace to
+// separate between instances of TimerStore. For instance, module "packages" would
+// use its module name for prefix.
+//
+// 2. TimerStore keeps the timers with a special prefix; it uses the timeout value
+// (block height/timestamp) as the key, so that the standard iterator would yield them
+// in the desired chronological order. TimeSStore also keeps the next-timeout values
+// (of block height/timestamp) to efficiently determine if the iterator is needed.
+// For instance, module "packages" may have two (block height) timeouts with their data
+// set to "first" and "second" respectively:
+//
+//     prefix: packages_Timer_Next_       key: BlockHeight      value: 150
+//     prefix: packages_Timer_Next_       key: BLockTimer       value: MaxInt64
+//     prefix: packages_Timer_Value_          key: 150          data: "first"
+//     prefix: packages_Entry_Value_          key: 180          data: "second"
+//
+// 3. TimerStore tracks the next-timeout for both block-height/block-timestamp. On
+// every call to Tick(), it tests the current ctx's block height/timestamp against the
+// respective next-timeout:
+//
+// 4. If the next-timeout is reached/passed, then it will iterate through the timer
```

Comment:
A thought: isn't it problematic that the callbacks might be invoked long after they were supposed to be called?
For example, say I have a callback that is supposed to be called in block 100 and the user ticks the timer only on block 1000.
Maybe there should be a global timer that is incremented on every new block so there will be a single reference of the current time to all timers

Reply:
The TimerStore is like a library, and users may use it as they see fit.
If exact timeouts are needed, the user can call AdvanceBlock() on every BeginBlock() of its own module.
If it is sufficient to have callbacks on epoch start/end, then the user can call AdvanceBlock() on epoch start/end (and save some CPU not checking blocks in between).
As long as the callback timing is deterministic, and the user controls the "AdvanceBlock()" calls, things are fine.

Reply:
Ideally, TimerStore will be its own module, and will have its own BeginBlock(), and won't depend on the user code calling AdvanceBlock(). There can also be a parameter to set the frequency of checking. Currently left for future work.

Reply:
I think your comment about calling AdvanceBlock() on every BeginBlock() of the user's module for exact timeouts should be explicitly documented in the TimerStore documentation as a best/common practice (so a forgetful user, like me, won't forget to do it)